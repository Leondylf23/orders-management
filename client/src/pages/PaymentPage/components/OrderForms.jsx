import { useDispatch, connect } from 'react-redux';
import { createStructuredSelector } from 'reselect';
import { FormattedMessage } from 'react-intl';
import { useEffect, useState } from 'react';
import PropTypes from 'prop-types';
import LocalOfferIcon from '@mui/icons-material/LocalOffer';

import { numberWithPeriods } from '@utils/allUtils';
import { selectProductDetail } from '@pages/ProductDetail/selectors';
import { setUserInputs } from '../actions';
import { selectUserInputData } from '../selectors';

import classes from '../style.module.scss';

const ProductDetailComponent = ({ productData, inputtedData }) => {
  const dispatch = useDispatch();
  const orderFormsDefault = {
    phone: '',
    address: '',
  };

  const [orderForm, setOrderForm] = useState(orderFormsDefault);

  const dataOnChange = (data) => {
    setOrderForm((prevVal) => {
      dispatch(setUserInputs({ ...inputtedData, ...{ ...prevVal, ...data } }));

      return { ...prevVal, ...data };
    });
  };

  useEffect(() => {
    if (inputtedData?.orderForm) {
      setOrderForm(inputtedData?.orderForm);
    }
  }, []);

  return (
    <div className={classes.componentContainer}>
      <h2 className={classes.title}>{productData?.title}</h2>
      <h4 className={classes.pageTitle}>
        <FormattedMessage id="payment_prct_detail_page_title" />
      </h4>
      <div className={classes.formInputs}>
        <label className={classes.label} htmlFor="phone">
          Phone Number
        </label>
        <input
          id="phone"
          type="tel"
          className={classes.input}
          value={orderForm?.phone}
          onChange={(e) => dataOnChange({ phone: e.target.value })}
        />
        <label className={classes.label} htmlFor="address">
          Alamat
        </label>
        <textarea
          id="address"
          type="text"
          className={classes.input}
          data-type="area"
          value={orderForm?.address}
          onChange={(e) => dataOnChange({ address: e.target.value })}
        />
      </div>
      <div className={classes.footer}>
        <h4 className={classes.footerTitle}>
          <FormattedMessage id="payment_total" />
        </h4>
        <div className={classes.priceContainer}>
          <LocalOfferIcon className={classes.icon} />
          <p className={classes.priceValue}>Rp. {numberWithPeriods(inputtedData?.totalPayment)}</p>
        </div>
      </div>
    </div>
  );
};

ProductDetailComponent.propTypes = {
  productData: PropTypes.object,
  inputtedData: PropTypes.object,
};

const mapStateToProps = createStructuredSelector({
  productData: selectProductDetail,
  inputtedData: selectUserInputData,
});

export default connect(mapStateToProps)(ProductDetailComponent);
