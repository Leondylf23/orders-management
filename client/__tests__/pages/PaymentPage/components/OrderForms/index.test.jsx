import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
// import RouterDom from 'react-router-dom';

import OrderForms from '@pages/PaymentPage/components/OrderForms';
import store from '@store';
import Language from '@containers/Language';

jest.mock('react-redux', () => ({
  ...jest.requireActual('react-redux'),
}));

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useNavigate: jest.fn(),
}));

jest.mock('reselect', () => ({
  ...jest.requireActual('reselect'),
}));

const ParentComponent = (children) => (
  <Provider store={store}>
    <Language>{children}</Language>
  </Provider>
);

// const navigate = jest.fn().mockName('navigate');

describe('Order Forms', () => {
  // beforeEach(() => {
  //     jest.spyOn(RouterDom, 'useNavigate').mockImplementation(() => navigate);
  // });

  test('Rendered', () => {
    const { getByTestId } = render(ParentComponent(<OrderForms />));
    const orderForms = getByTestId('payment-page-order-forms');
    expect(orderForms).toBeInTheDocument();
  });

  test('Should match with snapshot', () => {
    const orderForms = render(ParentComponent(<OrderForms />));
    expect(orderForms).toMatchSnapshot();
  });
});
