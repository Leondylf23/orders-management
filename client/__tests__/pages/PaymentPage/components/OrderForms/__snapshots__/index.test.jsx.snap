// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Order Forms Should match with snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="componentContainer"
        data-testid="payment-page-order-forms"
      >
        <h2
          class="title"
        />
        <h4
          class="pageTitle"
        >
          Isi form pengiriman di bawah ini
        </h4>
        <div
          class="formInputs"
        >
          <label
            class="label"
            for="phone"
          >
            Nomor Telepon
          </label>
          <input
            class="input"
            id="phone"
            type="tel"
            value=""
          />
          <label
            class="label"
            for="address"
          >
            Alamat Pengiriman
          </label>
          <textarea
            class="input"
            data-type="area"
            id="address"
            type="text"
          />
        </div>
        <div
          class="footer"
        >
          <h4
            class="footerTitle"
          >
            Total
          </h4>
          <div
            class="priceContainer"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium icon css-i4bv87-MuiSvgIcon-root"
              data-testid="LocalOfferIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="m21.41 11.58-9-9C12.05 2.22 11.55 2 11 2H4c-1.1 0-2 .9-2 2v7c0 .55.22 1.05.59 1.42l9 9c.36.36.86.58 1.41.58.55 0 1.05-.22 1.41-.59l7-7c.37-.36.59-.86.59-1.41 0-.55-.23-1.06-.59-1.42zM5.5 7C4.67 7 4 6.33 4 5.5S4.67 4 5.5 4 7 4.67 7 5.5 6.33 7 5.5 7z"
              />
            </svg>
            <p
              class="priceValue"
            >
              Rp. 
              0
            </p>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="componentContainer"
      data-testid="payment-page-order-forms"
    >
      <h2
        class="title"
      />
      <h4
        class="pageTitle"
      >
        Isi form pengiriman di bawah ini
      </h4>
      <div
        class="formInputs"
      >
        <label
          class="label"
          for="phone"
        >
          Nomor Telepon
        </label>
        <input
          class="input"
          id="phone"
          type="tel"
          value=""
        />
        <label
          class="label"
          for="address"
        >
          Alamat Pengiriman
        </label>
        <textarea
          class="input"
          data-type="area"
          id="address"
          type="text"
        />
      </div>
      <div
        class="footer"
      >
        <h4
          class="footerTitle"
        >
          Total
        </h4>
        <div
          class="priceContainer"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium icon css-i4bv87-MuiSvgIcon-root"
            data-testid="LocalOfferIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="m21.41 11.58-9-9C12.05 2.22 11.55 2 11 2H4c-1.1 0-2 .9-2 2v7c0 .55.22 1.05.59 1.42l9 9c.36.36.86.58 1.41.58.55 0 1.05-.22 1.41-.59l7-7c.37-.36.59-.86.59-1.41 0-.55-.23-1.06-.59-1.42zM5.5 7C4.67 7 4 6.33 4 5.5S4.67 4 5.5 4 7 4.67 7 5.5 6.33 7 5.5 7z"
            />
          </svg>
          <p
            class="priceValue"
          >
            Rp. 
            0
          </p>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
